veracruz-utils/src/csr.rs

```diff
@@ -0,0 +1,293 @@
+        0x02, 0x21, 0x00, 0xa0, 0xe0, 0x27, 0xf7, 0x23, 0x13, 0x92, 0x7f, 0x2e, 0x74, 0xcb, 0xf4, 0xf2,
+        0x57, 0x21, 0x1e, 0x53, 0x01, 0x63, 0xc0, 0xfc, 0xae, 0x44, 0x0c, 0x4f, 0x75, 0xf0, 0x1e, 0x00,
+        0x8f, 0xdd, 0x26, 0x02, 0x20, 0x24, 0xb5, 0xa8, 0x7a, 0x70, 0x04, 0x04, 0x3f, 0xfc, 0x09, 0xca,
+        0x04, 0x1d, 0xb5, 0x3b, 0x56, 0x89, 0x04, 0x7a, 0xda, 0x3e, 0x12, 0x50, 0x81, 0x2a, 0xa0, 0x3c,
+        0x5e, 0x29, 0xb6, 0xbb, 0xd1
+    ],
+    valid_from_location: (101, 101 + 13),
+    valid_until_location: (116, 116 + 13),
+    public_key_location: ( 274, 274 + 65),
+    signature_location: (430, 430 + 71),
+    signature_range: (4, 415),
+    overall_length_field_location: (2, 4),
+    overall_length_initial_value: 430 - 4,
+    signature_length_field_location: (428, 429),
+    signature_length_initial_value: 1,
+    enclave_hash_location: (383, 383 + 32),
+};
+
+pub fn generate_csr(template: &CsrTemplate, private_key: &EcdsaKeyPair) -> Result<Vec<u8>, CertError> {
+    let public_key = private_key.public_key().as_ref().clone();
+    let mut constructed_csr = template.template.to_vec();
+    if public_key.len() != (template.public_key_location.1 - template.public_key_location.0) {
+        return Err(CertError::InvalidLength { variable: "public_key", expected: template.public_key_location.1 - template.public_key_location.0, received: public_key.len() } );
+    }
+    constructed_csr.splice(
+        template.public_key_location.0..template.public_key_location.1,
+        public_key.iter().cloned(),
+    );
+
+    let rng = SystemRandom::new();
+    let signature: Vec<u8> = private_key.sign(&rng, &constructed_csr[template.signature_range.0..template.signature_range.1]).unwrap().as_ref().to_vec();
+
+
+    let signature_length = signature.len();
+    constructed_csr.splice(
+        template.signature_location.0..template.signature_location.1,
+        signature,
+    );
+
+    let signature_field_length:u8 = (template.signature_length_initial_value + signature_length as u8) as u8;
+    constructed_csr[template.signature_length_field_location.0] = signature_field_length;
+
+    let overall_length:u16 = (template.overall_length_initial_value + signature_length as u16) as u16;
+    if overall_length < 256 {
+        constructed_csr[template.overall_length_field_location.0] = overall_length as u8;
+    } else {
+        constructed_csr[template.overall_length_field_location.0] = ((overall_length & 0xff00) >> 8) as u8;
+        constructed_csr[template.overall_length_field_location.0 + 1] = (overall_length & 0xff) as u8;
+    }
+
+    return Ok(constructed_csr.clone());
+}
+
+const CSR_PUBKEY_LOCATION: (usize, usize) = (129 + 26, 220);
+pub fn convert_csr_to_cert(csr: &[u8], cert_template: &CertTemplate, enclave_hash: &[u8], private_key: &EcdsaKeyPair) -> Result<std::vec::Vec<u8>, CertError> {
+    let mut constructed_cert = cert_template.template.to_vec();
+    let valid_from = generate_utc_time(2021, 5, 2, 17, 1, 0)?;
```

Comment:
Should these dates be pulled out as constants, too?

Reply:
will do